sql: set fraction stale rows to 4 on sql stats tables

dhartunian opened the pull request:
Previously, stats collection on StatementStatisticsTableSchema and
TransactionStatisticsTableSchema were triggered at 20% stale rows
which was very low and would trigger often during flushing of SQL
statement stats.

This commit adds a migration that sets
`sql_stats_automatic_collection_fraction_stale_rows` at 4 which will
reduce stats collection overhead. Since this table is updated very
regularly, we expect stats to stay up to date even with the higher
threshold.

As recommended, we also set
`sql_stats_automatic_partial_collection_fraction_stale_rows` to 1.0

Michael Erickson contributed helpful insights:

See
https://github.com/cockroachdb/cockroach/issues/147443#issuecomment-2932501220
for calculations that led to the number selection.

There was an additional discussion on this PR regarding setting static
stats, that was also rejected since the queries running on these
tables are varied and complex.
See https://github.com/cockroachdb/cockroach/pull/151502#pullrequestreview-3125189444

Resolves: https://github.com/cockroachdb/cockroach/issues/147443
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151502)
<!-- Reviewable:end -->

alyshanjahani-crl reviewed (approved):
Bumping the storage parameter `sql_stats_automatic_collection_fraction_stale_rows` on these tables makes sense to me.

I wonder what the impact of stale table stats would be for this.. I imagine the query plans  for these tables don't change that often to begin with.

Regardless, at any non trivial workload the 90% fraction should trigger frequently for these tables.

One alternative would be to have a specific cluster setting that controls the storage parameter on these tables. But I'm happy with the way things are here.

RaduBerinde commented:
IMO we should disable auto stats on all system tables (and inject fixed stats as needed to get good plans for the queries we care about). Files https://github.com/cockroachdb/cockroach/issues/151587

dhartunian commented:
@RaduBerinde thanks for taking a look. is there an example I could follow to apply fixed stats to these tables?

rafiss commented:
I believe @RaduBerinde is referring to optimizer table stats for system tables, but IIUC, this PR is addressing behavior with o11y statement and transaction stats. This is a common source of confusion (https://github.com/cockroachdb/cockroach/issues/80279).

RaduBerinde commented:
@dhartunian here's an example: https://github.com/cockroachdb/cockroach/blob/e5d539416c469d9da48c1873600e3ca8dd7059c2/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_cascade#L259

yuzefovich commented:
We have `stats.DisallowedOnSystemTable` which determines whether a given system table ID can have auto stats collection on _and_ whether existing stats on that table can be used by the stats cache. The latter is controlled via `statsUsageAllowed` (which calls `DisallowedOnSystemTable`). I think we'll need to separate two concerns of this exported method into two separate helpers.

michae2 commented on pkg/sql/catalog/systemschema/system.go:
```diff
@@ -669,7 +669,7 @@ CREATE TABLE system.statement_statistics (
 			p99_latency
 		)
-)
+) WITH (sql_stats_automatic_collection_fraction_stale_rows = 0.9)
```
It's fine for this number to go above 1.0. Based on some [back-of-the-envelope math](https://github.com/cockroachdb/cockroach/issues/147443#issuecomment-2932501220) I was thinking 4.0, but if we want to be more conservative, that's fine. It might be nice to pick an even multiple of 0.2 (the default).

Whatever we pick, we should also set `sql_stats_automatic_partial_collection_fraction_stale_rows` to 1/4 of it.

michae2 reviewed (approved):
LGTM

Regarding whether to disable auto stats on these tables, a couple observations:
- `system.statement_statistics` has 8 secondary indexes
- `system.transaction_statistics` has 7 secondary indexes
- There are somewhat complex queries against both, for example:
   - https://github.com/cockroachdb/cockroach/blob/d5e306338eebd3ad77f1a9dcc124fcda30b619f9/pkg/sql/sql_activity_update_job.go#L726-L758
   - https://github.com/cockroachdb/cockroach/blob/d5e306338eebd3ad77f1a9dcc124fcda30b619f9/pkg/sql/sql_activity_update_job.go#L776-L852
 
For these reasons, I probably wouldn't disable auto stats on these tables. The tables and queries are complex enough that we probably want to give the optimizer up-to-date information.

If changing the setting doesn't help, we can always revisit the decision.

dhartunian committed:
sql: set fraction stale rows to 4 on sql stats tables

Previously, stats collection on StatementStatisticsTableSchema and
TransactionStatisticsTableSchema were triggered at 20% stale rows
which was very low and would trigger often during flushing of SQL
statement stats.

This commit adds a migration that sets
`sql_stats_automatic_collection_fraction_stale_rows` at 4 which will
reduce stats collection overhead. Since this table is updated very
regularly, we expect stats to stay up to date even with the higher
threshold.

As recommended, we also set
`sql_stats_automatic_partial_collection_fraction_stale_rows` to 1.0

Michael Erickson contributed helpful insights:

See
https://github.com/cockroachdb/cockroach/issues/147443#issuecomment-2932501220
for calculations that led to the number selection.

There was an additional discussion on this PR regarding setting static
stats, that was also rejected since the queries running on these
tables are varied and complex.
See https://github.com/cockroachdb/cockroach/pull/151502#pullrequestreview-3125189444

Resolves: #147443
Epic: None
Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> You're welcome!

> @michae2 reviewed 5 of 11 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151502#-:-OZpi-mX6G7sL6blqHiq:bnse8s5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, and @kyle-a-wong)



<!-- Sent from Reviewable.io -->
